RHDEVDOCS-4719+4724 Improve Tekton Hub config UX + Support for Crunchy Postgres with Tekton Hub (#54522)
openshift/openshift-docs: modules/op-migrating-tekton-hub-data-to-an-existing-crunchy-postgres-database.adoc

// This module is included in the following assembly:
//
// *cicd/pipelines/using-tekton-hub-with-openshift-pipelines.adoc

:_content-type: PROCEDURE
[id="migrating-tekton-hub-data-to-an-existing-crunchy-postgres-database_{context}"]
= Optional: Migrating {tekton-hub} data to an existing Crunchy Postgres database

[role="_abstract"]

> If a Crunchy Postgres database pod already exists, cluster administrators can migrate {tekton-hub} data from the internal or default database to the external Crunchy Postgres database.

Reviewer: ```suggestion
If  Tekton Hub is already installed with default database and now if you want to use your own database(e.g. Crunchy Postgres), cluster administrators can migrate {tekton-hub} data from the internal or default database to the external Crunchy Postgres database.
```

> {tekton-hub} supports use of Crunchy Postgres as a custom database. For a pre-installed {tekton-hub} with default database, cluster administrators can use Crunchy Postgres as a custom database after migrating the {tekton-hub} data from the internal or default database to the external Crunchy Postgres database.

Reviewer 2: ```suggestion
{tekton-hub} supports the use of Crunchy Postgres as a custom database. For a pre-installed {tekton-hub} with default database, cluster administrators can use Crunchy Postgres as a custom database after migrating the {tekton-hub} data from the internal or default database to the external Crunchy Postgres database.
```

[discrete]
.Procedure

. Dump the existing data from the internal or default database into a file in the pod.
+
.Example: Dump data
[source,terminal]
----
$ pg_dump -Ft -h localhost -U postgres hub -f /tmp/hub.dump
----

. Copy the file containing the data dump to your local system.
+
.Command format
[source,terminal]
----
$ oc cp -n <namespace> <podName>:<path-to-hub.dump> <path-to-local-system>
----
.Example
[source,terminal]
----
$ oc cp -n openshift-pipelines tekton-hub-db-7d6d888c67-p7mdr:/tmp/hub.dump /home/test_user/Downloads/hub.dump
----

> . Copy the file containing data dump from the local system to the pod running the external Crunchy Postgres database.

Reviewer 2: ```suggestion
. Copy the file containing the data dump from the local system to the pod running the external Crunchy Postgres database.
```
```suggestion
. Copy the file that contains the data dump from the local system to the pod running the external Crunchy Postgres database.
```

psql (14.4)
Type "help" for help.
----

. Find the `pg_hba.conf` file.
+
[source,terminal]
----
postgres=# SHOW hba_file;
         hba_file         
--------------------------
 /pgdata/pg14/pg_hba.conf
(1 row)

postgres=#
----

. Exit from the database.

. Check if the `pg_hba.conf` file has the entry `host all all 0.0.0.0/0 md5`, which is necessary for accessing all incoming connections. If necessary, add the entry at the end of the `pg_hba.conf` file.
+
.Example: `pg_hba.conf` file
[source,terminal]
----
sh-4.4$ cat /pgdata/pg14/pg_hba.conf

# Do not edit this file manually!
# It will be overwritten by Patroni!
local all "postgres" peer
hostssl replication "_crunchyrepl" all cert
hostssl "postgres" "_crunchyrepl" all cert
host all "_crunchyrepl" all reject
hostssl all all all md5
host  all  all 0.0.0.0/0 md5
----

. Save the `pg_hba.conf` file and reload the database.
+
[source,terminal]
----
sh-4.4$ psql -U postgres
psql (14.4)
Type "help" for help.

postgres=# SHOW hba_file;
         hba_file         
--------------------------
 /pgdata/pg14/pg_hba.conf
(1 row)

postgres=# SELECT pg_reload_conf();
 pg_reload_conf
----------------
 t
(1 row)
----

. Exit the database.

> . Verify that a secret named `tekton-hub-db` in the target namespace exixts with the following keys: 

Reviewer 2: ```suggestion
. Verify that a secret named `tekton-hub-db` in the target namespace exists with the following keys: 
```
```suggestion
. Verify that a secret named `tekton-hub-db` in the target namespace has the following keys: 
```

    app: tekton-hub-db
type: Opaque
stringData:
  POSTGRES_HOST: test-primary.openshift-operators.svc 
  POSTGRES_DB: test
  POSTGRES_USER: test
  POSTGRES_PASSWORD: woXOisU5>ocJiTF7y{{;1[Q( 
  POSTGRES_PORT: '5432'
...
----
+
[NOTE]
====
The value of the `POSTGRES_HOST` field is encoded as a secret. You can decode the value of the Crunchy Postgres host by using the following example.
+
.Example: Decode the secret value of a Crunchy Postgres host 
[source,terminal]
----
$ echo 'aGlwcG8tcHJpbWFyeS5vcGVuc2hpZnQtb3BlcmF0b3JzLnN2YyA=' | base64 --decode
test-primary.openshift-operators.svc
----
====

. Verify that in the `TektonHub` CR, the value of the database secret attribute is `tekton-hub-db`.
+
.Example: TektonHub CR with the name of the database secret 
[source,yaml]
----
apiVersion: operator.tekton.dev/v1alpha1
kind: TektonHub
metadata:
  name: hub
spec:
  targetNamespace: openshift-pipelines
  db:                      
    secret: tekton-hub-db 
...
----

. To associate the external Crunchy Postgres database with {tekton-hub}, replace any existing `TektonHub` CR with the updated `TektonHub` CR.
+
[source,terminal]
----
$ oc replace -f <updated-tekton-hub-cr>.yaml
----

> . Check the status of the {tekton-hub}. The updated `TektonHub` CR might take some time to attain steady state.

Reviewer 2: ```suggestion
. Check the status of the {tekton-hub}. The updated `TektonHub` CR might take some time to attain a steady state.
```

Reviewer 2: Multiple instances in these files.
